File: mRemoteV1.ConvertedToC#/Config/Putty/RegistryProvider.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;
using System.Management;
using mRemoteNG.App;
using mRemoteNG.Messages;
using Microsoft.Win32;
using mRemoteNG.Connection.Protocol;
using System.Security.Principal;

namespace mRemoteNG.Config.Putty
{
	public class RegistryProvider : Provider
	{

		#region "Public Methods"
		public override string[] GetSessionNames(bool raw = false)
		{
			RegistryKey sessionsKey = Registry.CurrentUser.OpenSubKey(PuttySessionsKey);
			if (sessionsKey == null)
				return new string[];

			List<string> sessionNames = new List<string>();
			foreach (string sessionName in sessionsKey.GetSubKeyNames()) {
				if (raw) {
					sessionNames.Add(sessionName);
				} else {
					sessionNames.Add(System.Web.HttpUtility.UrlDecode(sessionName.Replace("+", "%2B")));
				}
			}

			if (raw) {
				// Do not localize
				if (!sessionNames.Contains("Default%20Settings")) {
					sessionNames.Insert(0, "Default%20Settings");
				}
			} else {
				if (!sessionNames.Contains("Default Settings")) {
					sessionNames.Insert(0, "Default Settings");
				}
			}

			return sessionNames.ToArray();
		}

		public override Connection.PuttySession.Info GetSession(string sessionName)
		{
			RegistryKey sessionsKey = Registry.CurrentUser.OpenSubKey(PuttySessionsKey);
			if (sessionsKey == null)
				return null;

			RegistryKey sessionKey = sessionsKey.OpenSubKey(sessionName);
			if (sessionKey == null)
				return null;

			sessionName = System.Web.HttpUtility.UrlDecode(sessionName.Replace("+", "%2B"));

			Connection.PuttySession.Info sessionInfo = new Connection.PuttySession.Info();
			var _with1 = sessionInfo;
			_with1.PuttySession = sessionName;
			_with1.Name = sessionName;
			_with1.Hostname = sessionKey.GetValue("HostName");
			_with1.Username = sessionKey.GetValue("UserName");
			string protocol = sessionKey.GetValue("Protocol");
			if (protocol == null)
				protocol = "ssh";
			switch (protocol.ToLowerInvariant()) {
				case "raw":
					_with1.Protocol = Protocols.RAW;
					break;
				case "rlogin":
					_with1.Protocol = Protocols.Rlogin;
					break;
				case "serial":
					return null;
				case "ssh":
					object sshVersionObject = sessionKey.GetValue("SshProt");
					if (sshVersionObject != null) {
						int sshVersion = Convert.ToInt32(sshVersionObject);
						if (sshVersion >= 2) {
							_with1.Protocol = Protocols.SSH2;
						} else {
							_with1.Protocol = Protocols.SSH1;
						}
					} else {
						_with1.Protocol = Protocols.SSH2;
					}
					break;
				case "telnet":
					_with1.Protocol = Protocols.Telnet;
					break;
				default:
					return null;
			}
			_with1.Port = sessionKey.GetValue("PortNumber");

			return sessionInfo;
		}

		public override void StartWatcher()
		{
			if (_eventWatcher != null)
				return;

			try {
				string currentUserSid = WindowsIdentity.GetCurrent().User.Value;
				string key = string.Join("\\", {
					currentUserSid,
					PuttySessionsKey
				}).Replace("\\", "\\\\");
				WqlEventQuery query = new WqlEventQuery(string.Format("SELECT * FROM RegistryTreeChangeEvent WHERE Hive = 'HKEY_USERS' AND RootPath = '{0}'", key));
				_eventWatcher = new ManagementEventWatcher(query);
				_eventWatcher.EventArrived += OnManagementEventArrived;
				_eventWatcher.Start();
			} catch (Exception ex) {
				Runtime.MessageCollector.AddExceptionMessage("PuttySessions.Watcher.StartWatching() failed.", ex, MessageClass.WarningMsg, true);
			}
		}

		public override void StopWatcher()
		{
			if (_eventWatcher == null)
				return;
			_eventWatcher.Stop();
			_eventWatcher.Dispose();
			_eventWatcher = null;
		}
		#endregion

		#region "Private Fields"
		private const string PuttySessionsKey = "Software\\SimonTatham\\PuTTY\\Sessions";
			#endregion
		private static ManagementEventWatcher _eventWatcher;

		#region "Private Methods"
		private void OnManagementEventArrived(object sender, EventArrivedEventArgs e)
		{
			OnSessionChanged(new SessionChangedEventArgs());
		}
		#endregion
	}
}

File: mRemoteV1/Config/Putty/RegistryProvider.vb
Imports System.Management
Imports System.Security.Principal
Imports System.Web
Imports Microsoft.Win32
Imports mRemote3G.App
Imports mRemote3G.Connection.Protocol
Imports mRemote3G.Connection.PuttySession
Imports mRemote3G.Messages

Namespace Config.Putty
    Public Class RegistryProvider
        Inherits Provider

#Region "Public Methods"

        Public Overrides Function GetSessionNames(Optional ByVal raw As Boolean = False) As String()
            Dim sessionsKey As RegistryKey = Registry.CurrentUser.OpenSubKey(PuttySessionsKey)
            If sessionsKey Is Nothing Then Return New String() {}

            Dim sessionNames As New List(Of String)
            For Each sessionName As String In sessionsKey.GetSubKeyNames()
                If raw Then
                    sessionNames.Add(sessionName)
                Else
                    sessionNames.Add(HttpUtility.UrlDecode(sessionName.Replace("+", "%2B")))
                End If
            Next

            If raw Then
                If Not sessionNames.Contains("Default%20Settings") Then ' Do not localize
                    sessionNames.Insert(0, "Default%20Settings")
                End If
            Else
                If Not sessionNames.Contains("Default Settings") Then
                    sessionNames.Insert(0, "Default Settings")
                End If
            End If

            Return sessionNames.ToArray()
        End Function

        Public Overrides Function GetSession(sessionName As String) As PuttyInfo
            Dim sessionsKey As RegistryKey = Registry.CurrentUser.OpenSubKey(PuttySessionsKey)
            If sessionsKey Is Nothing Then Return Nothing

            Dim sessionKey As RegistryKey = sessionsKey.OpenSubKey(sessionName)
            If sessionKey Is Nothing Then Return Nothing

            sessionName = HttpUtility.UrlDecode(sessionName.Replace("+", "%2B"))

            Dim sessionInfo As New PuttyInfo
            With sessionInfo
                .PuttySession = sessionName
                .Name = sessionName
                .Hostname = sessionKey.GetValue("HostName")
                .Username = sessionKey.GetValue("UserName")
                Dim protocol As String = sessionKey.GetValue("Protocol")
                If protocol Is Nothing Then protocol = "ssh"
                Select Case protocol.ToLowerInvariant()
                    Case "raw"
                        .Protocol = Protocols.RAW
                    Case "rlogin"
                        .Protocol = Protocols.Rlogin
                    Case "serial"
                        Return Nothing
                    Case "ssh"
                        Dim sshVersionObject As Object = sessionKey.GetValue("SshProt")
                        If sshVersionObject IsNot Nothing Then
                            Dim sshVersion = CType(sshVersionObject, Integer)
                            If sshVersion >= 2 Then
                                .Protocol = Protocols.SSH2
                            Else
                                .Protocol = Protocols.SSH1
                            End If
                        Else
                            .Protocol = Protocols.SSH2
                        End If
                    Case "telnet"
                        .Protocol = Protocols.Telnet
                    Case Else
                        Return Nothing
                End Select
                .Port = sessionKey.GetValue("PortNumber")
            End With

            Return sessionInfo
        End Function

        Public Overrides Sub StartWatcher()
            If _eventWatcher IsNot Nothing Then Return

            Try
                Dim currentUserSid As String = WindowsIdentity.GetCurrent().User.Value
                Dim key As String = String.Join("\", {currentUserSid, PuttySessionsKey}).Replace("\", "\\")
                Dim _
                    query As _
                        New WqlEventQuery(
                            String.Format(
                                "SELECT * FROM RegistryTreeChangeEvent WHERE Hive = 'HKEY_USERS' AND RootPath = '{0}'",
                                key))
                _eventWatcher = New ManagementEventWatcher(query)
                AddHandler _eventWatcher.EventArrived, AddressOf OnManagementEventArrived
                _eventWatcher.Start()
            Catch ex As Exception
                Runtime.MessageCollector.AddExceptionMessage("PuttySessions.Watcher.StartWatching() failed.", ex,
                                                             MessageClass.WarningMsg, True)
            End Try
        End Sub

        Public Overrides Sub StopWatcher()
            If _eventWatcher Is Nothing Then Return
            _eventWatcher.Stop()
            _eventWatcher.Dispose()
            _eventWatcher = Nothing
        End Sub

#End Region

#Region "Private Fields"

        Private Const PuttySessionsKey As String = "Software\SimonTatham\PuTTY\Sessions"
        Private Shared _eventWatcher As ManagementEventWatcher

#End Region

#Region "Private Methods"

        Private Sub OnManagementEventArrived(sender As Object, e As EventArrivedEventArgs)
            OnSessionChanged(New SessionChangedEventArgs())
        End Sub

#End Region
    End Class
End Namespace

